src/framebuffer/refresh.rs

```diff
@@ -126,34 +120,32 @@ impl<'a> framebuffer::FramebufferRefresh for core::Framebuffer<'a> {
             ..Default::default()
         };
 
-        let pt: *const mxcfb_update_data = &whole;
-        unsafe {
-            libc::ioctl(self.device.as_raw_fd(), common::MXCFB_SEND_UPDATE, pt);
+        let update_succeeded = if let Some(ref swtfb_client) = self.swtfb_client {
+            swtfb_client.send_mxcfb_update(&whole)
+        } else {
+            let pt: *const mxcfb_update_data = &whole;
+            (unsafe { libc::ioctl(self.device.as_raw_fd(), common::MXCFB_SEND_UPDATE, pt) }) > 0
+        };
+
+        if !update_succeeded {
+            warn!("Sending full_refresh update failed!")
         }
 
         match mode {
             PartialRefreshMode::Wait | PartialRefreshMode::DryRun => {
-                let mut markerdata = mxcfb_update_marker_data {
-                    update_marker: whole.update_marker,
-                    collision_test: 0,
-                };
-                unsafe {
-                    if libc::ioctl(
-                        self.device.as_raw_fd(),
-                        common::MXCFB_WAIT_FOR_UPDATE_COMPLETE,
-                        &mut markerdata,
-                    ) < 0
-                    {
-                        warn!("WAIT_FOR_UPDATE_COMPLETE failed after a partial_refresh(..)");
-                    }
-                }
-                markerdata.collision_test
+                self.wait_refresh_complete(whole.update_marker)
             }
             PartialRefreshMode::Async => whole.update_marker,
         }
     }
 
     fn wait_refresh_complete(&self, marker: u32) -> u32 {
+        if let Some(ref swtfb_client) = self.swtfb_client {
+            swtfb_client.wait_for_update_complete();
```

Comment:
```suggestion
            return swtfb_client.wait_for_update_complete();
```
and move the comment there under a TODO (assuming this could be achieved in the future?)

Reply:
This would cause a syntax error as `swtfb_client.wait_for_update_complete()` returns nothing. There is a lot of mostly copied c code that in the original client also didn't assume any error condition. Probably adding such to the function would be better first.

Reply:
ofc what I'm saying is having `wait_for_update_complete` return the error code so 
```rust
            // Assume success
            return 0;
```
is within `wait_for_update_complete`'s body i.e. geographically close to the logic.

Reply:
While I would agree in general, I disagree for this case. Rm2fb doesn't support the return type which is a collision check. Adding it within swtfb_client would suggest a feature that is doesn't have. The method is also slightly different phrased to match the ioctl name that rm2fb handles, not libremarkables one.

While the ioctl technicalla gets the marker data and "returns" the collision value (u32), I would rather the missing support to be obvious here (rm2fb doesn't care about the marker).